tap_wikipedia/wikipedia_streams.py

```diff
@@ -0,0 +1,127 @@
+                                "url", th.StringType),
+                            ),
+                            ),
+
+                th.Property("sublinks",
+                            th.ArrayType(
+                                th.ObjectType(
+                                    th.Property("anchor", th.StringType),
+                                    th.Property("link", th.StringType)
+                                )
+                            ),
+                            )
+            ).to_dict(),
+        )
+        self.__logger = logging.getLogger(__name__)
+        self.wikipedia_config = wikipedia_config
+
+    def __get_featured_articles(self) -> List[str]:
+        """Retrieve URLs of Featured Wikipedia Articles"""
+
+        url = "https://en.wikipedia.org/wiki/Wikipedia:Featured_articles"
+        links = []
+
+        response = requests.get(url)
+        soup = BeautifulSoup(response.text, "html.parser")
+
+        for link in soup.find_all('a'):
+            current_link = str(link.get('href'))
+            if current_link[:5] == "/wiki":
+                links.append(current_link)
+
+        featured_urls = list("https://en.wikipedia.org" +
+                             link for link in links)
+
+        return featured_urls
+
+    def __get_wikipedia_records(self, cached_file_path) -> List[Any]:
+        """Retrieve list of Wikipedia Records"""
+
+        # Setup parser
+        parser = sax.make_parser()
+        parser.setFeature(sax.handler.feature_namespaces, 0)
+
+        # Instantiate SAX Handler and run parser
+        Handler = WikipediaAbstractsParser()
+        parser.setContentHandler(Handler)
+        parser.parse(cached_file_path)
+
+        # Return list of records
+        return Handler.getRecords()
+
+    def get_records(
+        self,
+        context: Dict | None
+    ) -> Iterable[Dict]:
+        """Generate Stream of Wikipedia Records"""
+
+        # Set default config values
+        default_url = 'https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-abstract1.xml.gz'
+        default_cache_dir = user_cache_dir("abstracts", "tap-wikipedia")
+
+        # Get config values
+        url = self.wikipedia_config.get("abstracts-dump-url", default_url)
+        cache_dir = Path(self.wikipedia_config.get(
+            "cache-path", default_cache_dir))
+        subset_specification = self.wikipedia_config.get(
+            "subset-specification", [])
+
+        # Get cache directory
+        abstractsCache = FileCache(cache_dir_path=cache_dir)
+        try:
+            cached_file_path = abstractsCache.get_file(url)
+        except:
+            self.__logger.warning(
+                "error downloading Wikipedia dump", exc_info=True)
+            return 1
+
+        # get Wikipedia records
+        records = self.__get_wikipedia_records(cached_file_path)
+
+        # returns a list of featured Wikipedia Article records
+        def get_featured_records() -> List[Any]:
```

Comment:
Use generators / `yield` / `yield from` rather than returning lists or tuples if you don't need to access anything in the sequence by index. A generator qualifies as `Iterable`. That way you're not buffering everything in memory.

Reply:
I have switched to `yield` / `yield from` for the final records.